apps_download/models/product_product.py

```diff
@@ -0,0 +1,107 @@
+                        raise ValidationError(
+                            _('Error: You cannot create recursive dependency.')
+                        )
+                    product_dependent_ids += child
+                    child_dependency_check(product_dependent_ids,
+                                           child.dependent_product_ids)
+                return res
+
+            child_dependency_check(product, product.dependent_product_ids)
+
+    @api.multi
+    def create_dependency_list(self):
+        ret_val = {}
+
+        def child_dependency(children):
+            res = self.env['product.product']
+            for child in children:
+                if not child.dependent_product_ids:
+                    continue
+                res += child.dependent_product_ids
+                res += child_dependency(child.dependent_product_ids)
+            return res
+        for product in self:
+            ret_val[product.id] = product.dependent_product_ids
+            if product.dependent_product_ids:
+                ret_val[product.id] += child_dependency(
+                    product.dependent_product_ids)
+        return ret_val
+
+    @api.multi
+    def generate_zip_file(self):
+        for product in self:
+            if not product.module_path:
+                continue
+            tmpdir = tempfile.mkdtemp()
+            tmpdir2 = tempfile.mkdtemp()
+            dependent_product_ids = product.create_dependency_list(
+                )[product.id]
+            for dependent_product in dependent_product_ids:
+                if not dependent_product.module_path:
+                    continue
+                subprocess.Popen(['cp', '-r', dependent_product.module_path,
+                                  tmpdir], stdout=subprocess.PIPE)
+            subprocess.Popen(['cp', '-r', product.module_path, tmpdir],
+                             stdout=subprocess.PIPE)
+            time_value = time.strftime(
+                '_%y%m%d_%H%M%S')
+
+            tmpzipfile = os.path.join(tmpdir2, product.name) + time_value
+            shutil.make_archive(tmpzipfile, 'zip', tmpdir)
+            tmpzipfile = tmpzipfile + '.zip'
+            with open(tmpzipfile, "rb") as fileobj:
+                try:
+                    data_encode = base64.encodestring(fileobj.read())
+                    self.env['ir.attachment'].create({
+                        'datas': data_encode,
+                        'datas_fname': tmpzipfile,
+                        'type': 'binary',
+                        'name': product.name + time_value + '.zip',
+                        'res_model': product._name,
+                        'res_id': product.id,
+                        'product_downloadable': True,
+                    })
+                except:
+                    _logger.error('Error creating attachment %s' % tmpzipfile)
+            try:
+                shutil.rmtree(tmpdir)
+                shutil.rmtree(tmpdir2)
```

Comment:
If the first fails, this will not process. I recommend splitting into two try/excepts if you actually expect them to fail for some reason.

That said, why would these fail in normal conditions? IMO this is a true error that should not be masked, otherwise the file system will be filling up with temp directories.